sysroot/usr/include/kernel/lib/ring_buffer.hpp

```diff
@@ -13,84 +13,129 @@
 #pragma once
 
 #include <stdint.h>
+#include <lib/errno.h>
 
-typedef struct ring_buff {
-    uint8_t* data;
+template <typename T, size_t S>
+class RingBuffer {
+public:
+    /**
+     * @brief Initializes the circular buffer and allocates
+     * the data memory.
+     */
+    explicit RingBuffer()
+        : head(0)
+        , tail(0)
+        , length(0)
+    {
+        // Default constructor
+    }
+    /**
+     * @brief Writes a byte into the circular buffer.
+     *
+     * @param val Data to write to the buffer
+     */
+    int Enqueue(T val)
+    {
+        int status = 0;
+        // Check if the buffer is full. If so, we can't enqueue.
+        if (IsFull()) {
+            status = -1;
+            errno = ENOBUFS;
+        } else {
+            // Write the data at the write index
+            this->data[this->head] = val;
+            this->head = ((this->head + 1) % S);
+            ++this->length;
+        }
+        // Return the status code
+        return status;
+    }
+    /**
+     * @brief Removes a bytes from the circular buffer and writes
+     * the value of the byte to the data pointer.
+     *
+     * @param buf Buffer to contain the data
+     * @return int Returns 0 on success and -1 on error.
+     */
+    int Dequeue(T* buf)
```

Comment:
Can we maybe get an option to dequeue by value?

Reply:
Resolved in 04f2009.